### ISSUE TITLE
Increasing mktxp collection times/latency (specifically HealthCollector)

### ISSUE THREAD

**Issue by PhireSideZA** [akpw/mktxp#34]
Hi,

I seem to be having an issue with the HealthCollector latency increasing slowly over time. This is with one host monitored. I have set a script to restart the containers at 01:00, as I receive timeouts if the container is running for too long a time.

This is running in a container on an HP N40L Microserver, running OpenMediaVault (Debian Bullseye) and the host is a Mikrotik HeX S

<img width="1348" alt="Screenshot 2022-08-15 at 16 13 04" src="https://user-images.githubusercontent.com/66358625/184652120-946766d8-7aa5-4e8c-abcd-5b40a3d8ca18.png">

**Comment by M0r13n** [comment]
@akpw I also noticed this behavior. But for me it looks like this actually a bug in mktxp?

At this line: https://github.com/akpw/mktxp/blob/7aeb02ac7147a2f060f6493684c60053b44107b6/mktxp/flow/collector_handler.py#L37


The method looks like this:

```python
    def collect(self):
        yield from self.collector_registry.bandwidthCollector.collect()

        for router_entry in self.entries_handler.router_entries:
            if not router_entry.api_connection.is_connected():
                # let's pick up on things in the next run
                router_entry.api_connection.connect()
                continue

            for collector_ID, collect_func in self.collector_registry.registered_collectors.items():                
                start = default_timer()
                yield from collect_func(router_entry)
                router_entry.time_spent[collector_ID] += default_timer() - start
```

The `time_spent` for every router entry is always only increased, but never reduced. Or do I miss something?

**Comment by M0r13n** [comment]
Nevermind. The metric is a counter which is a scalar value that is only going to be incremented. Hence the `rate()` query in Prometheus...

**Comment by akpw** [comment]
Indeed. Regarding the HealthCollector issue, there was a recent [pull request](https://github.com/akpw/mktxp/pull/32) so pls feel free to try it out (installing from this repo, since the merge is not yet on PyPi). I will hopefully get to it in a week or two, publishing next version with some more fixes.

**Comment by M0r13n** [comment]
This is actually my PR :-D

But this doesn't fix this issue. I am running a patched version of mktxp, but am still getting this error. But I think that there is some kind of memory leak, as the CPU usage and memory consumption increases linearly over time:

![image](https://user-images.githubusercontent.com/31622033/190859875-eeca5411-3feb-4c56-aadb-4b1356fbe10e.png)

**Comment by M0r13n** [comment]
I found the issue: The problem is that a mutable object is passed as a default object. In this case a list is passed as a default object for `metric_labels` in the `BaseCollector` methods. If any function modifies the object (e.g. by appending an item to a list), the default value is in effect modified. 

See [this line for example](https://github.com/akpw/mktxp/blob/ad3f4b2f5d8b7d3e720d3ce5a9576a2ecfc6928a/mktxp/collector/base_collector.py#L44). [This method](https://github.com/akpw/mktxp/blob/ad3f4b2f5d8b7d3e720d3ce5a9576a2ecfc6928a/mktxp/collector/base_collector.py#L57)  appends to `metric_labels`. Thus the length of `metric_labels` grows forever until is reaches a size of hundreds of thousands of entries.

I will a submit PR tomorrow.

**Comment by akpw** [comment]
@M0r13n thanks much for investigating on this one! I went fwd and submitted a quick fix, please review.  I should hopefully have some time to publish a bug fix release tomorrow, and this one is definitely important to have there.

**Comment by M0r13n** [comment]
That looks great. Exactly what I would have done. Thank you for this quick fix. 👍🏻

**Comment by akpw** [comment]
should be fixed in [0.35](https://pypi.org/project/mktxp/)

**Comment by akpw** [comment]
closing for now, pls feel free to check & reopen if needed
